project-base/storefront/hooks/productLists/comparison/useComparison.tsx

```diff
@@ -0,0 +1,78 @@
+import {
+    ProductListTypeEnumApi,
+    useAddProductToComparisonMutationApi,
+    useCleanProductListMutationApi,
+    useComparisonQueryApi,
+    useRemoveProductFromComparisonMutationApi,
+} from 'graphql/generated';
+import { showErrorMessage, showSuccessMessage } from 'helpers/toasts';
+import { useIsUserLoggedIn } from 'hooks/auth/useIsUserLoggedIn';
+import { useProductList } from 'hooks/productLists/useProductList';
+import useTranslation from 'next-translate/useTranslation';
+import { useEffect, useState } from 'react';
+import { usePersistStore } from 'store/usePersistStore';
+
+export const useComparison = () => {
+    const { t } = useTranslation();
+    const isUserLoggedIn = useIsUserLoggedIn();
+
+    const [, addProductToListMutation] = useAddProductToComparisonMutationApi();
+    const [, removeProductFromListMutation] = useRemoveProductFromComparisonMutationApi();
+    const [, cleanListMutation] = useCleanProductListMutationApi();
+
+    const comparisonUuid = usePersistStore((store) => store.comparisonUuid);
+    const updateComparisonUuid = usePersistStore((store) => store.updateComparisonUuid);
+    const [isPopupCompareOpen, setIsPopupCompareOpen] = useState(false);
+
+    const [{ data: comparisonData, fetching }] = useComparisonQueryApi({
+        variables: { input: { uuid: comparisonUuid, type: ProductListTypeEnumApi.ComparisonApi } },
+        pause: !comparisonUuid && !isUserLoggedIn,
+    });
+
+    useEffect(() => {
+        if (comparisonData?.productList?.uuid) {
+            updateComparisonUuid(comparisonData.productList.uuid);
+        }
+    }, [comparisonData?.productList?.uuid]);
+
+    const { cleanList, isProductInList, toggleProductInList } = useProductList(
+        ProductListTypeEnumApi.ComparisonApi,
+        comparisonUuid,
+        comparisonData,
+        {
+            cleanList: cleanListMutation,
+            removeProductFromList: removeProductFromListMutation,
+            addProductToList: addProductToListMutation,
```

Comment:
The same topic which I've pointed out in previous comments. This is just breaking DRY principle. This can be implemented into `useProductList`. Since it's the same on both Wishlist and Comparison.

Reply:
As discussed, I agree with you. After going through it once again, I moved the layer of generic code up, now making basically everything generic. You can check [the commit, in which I do it](https://github.com/shopsys/shopsys/pull/2901/commits/40f0a750f24c4d5c83d4bfc417cf203a2727053d) and the modified upgrade notes.